File: packages/shared/src/protocol.ts
import type { UIElement, DeviceInfo, InstalledApp } from "./types.js";

export type DeviceMessage =
  | { type: "auth"; apiKey: string; deviceInfo?: DeviceInfo }
  | { type: "screen"; requestId: string; elements: UIElement[]; screenshot?: string; packageName?: string }
  | { type: "result"; requestId: string; success: boolean; error?: string; data?: string }
  | { type: "goal"; text: string }
  | { type: "pong" }
  | { type: "heartbeat"; batteryLevel: number; isCharging: boolean }
  | { type: "apps"; apps: InstalledApp[] }
  | { type: "stop_goal" }
  // Voice overlay
  | { type: "voice_start" }
  | { type: "voice_chunk"; data: string }
  | { type: "voice_stop"; action: "send" | "cancel" };

export type ServerToDeviceMessage =
  | { type: "auth_ok"; deviceId: string }
  | { type: "auth_error"; message: string }
  | { type: "get_screen"; requestId: string }
  | { type: "tap"; requestId: string; x: number; y: number }
  | { type: "type"; requestId: string; text: string }
  | { type: "swipe"; requestId: string; x1: number; y1: number; x2: number; y2: number; duration?: number }
  | { type: "enter"; requestId: string }
  | { type: "back"; requestId: string }
  | { type: "home"; requestId: string }
  | { type: "longpress"; requestId: string; x: number; y: number }
  | { type: "launch"; requestId: string; packageName: string; intentUri?: string; intentExtras?: Record<string, string> }
  | { type: "clear"; requestId: string }
  | { type: "clipboard_set"; requestId: string; text: string }
  | { type: "clipboard_get"; requestId: string }
  | { type: "paste"; requestId: string }
  | { type: "open_url"; requestId: string; url: string }
  | { type: "switch_app"; requestId: string; packageName: string }
  | { type: "notifications"; requestId: string }
  | { type: "keyevent"; requestId: string; code: number }
  | { type: "open_settings"; requestId: string; setting?: string }
  | { type: "wait"; requestId: string; duration?: number }
  | { type: "intent"; requestId: string; intentAction: string; intentUri?: string; intentType?: string; intentExtras?: Record<string, string>; packageName?: string }
  | { type: "ping" }
  | { type: "goal_started"; sessionId: string; goal: string }
  | { type: "goal_completed"; sessionId: string; success: boolean; stepsUsed: number }
  // Voice overlay
  | { type: "transcript_partial"; text: string }
  | { type: "transcript_final"; text: string };

export type DashboardMessage =
  | { type: "device_online"; deviceId: string; name: string }
  | { type: "device_offline"; deviceId: string }
  | { type: "step"; sessionId: string; step: number; action: Record<string, unknown>; reasoning: string; screenHash: string }
  | { type: "goal_started"; sessionId: string; goal: string; deviceId: string }
  | { type: "goal_completed"; sessionId: string; success: boolean; stepsUsed: number }
  | { type: "device_status"; deviceId: string; batteryLevel: number; isCharging: boolean };

File: android/app/src/main/java/com/thisux/droidclaw/model/Protocol.kt
package com.thisux.droidclaw.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

@Serializable
data class AuthMessage(
    val type: String = "auth",
    val apiKey: String,
    val deviceInfo: DeviceInfoMsg? = null
)

@Serializable
data class DeviceInfoMsg(
    val model: String,
    val manufacturer: String,
    val androidVersion: String,
    val screenWidth: Int,
    val screenHeight: Int,
    val batteryLevel: Int,
    val isCharging: Boolean
)

@Serializable
data class ScreenResponse(
    val type: String = "screen",
    val requestId: String,
    val elements: List<UIElement>,
    val screenHash: String? = null,
    val screenshot: String? = null,
    val packageName: String? = null
)

@Serializable
data class ResultResponse(
    val type: String = "result",
    val requestId: String,
    val success: Boolean,
    val error: String? = null,
    val data: String? = null
)

@Serializable
data class GoalMessage(
    val type: String = "goal",
    val text: String
)

@Serializable
data class PongMessage(
    val type: String = "pong"
)

@Serializable
data class HeartbeatMessage(
    val type: String = "heartbeat",
    val batteryLevel: Int,
    val isCharging: Boolean
)

@Serializable
data class InstalledAppInfo(
    val packageName: String,
    val label: String,
    val intents: List<String> = emptyList()
)

@Serializable
data class AppsMessage(
    val type: String = "apps",
    val apps: List<InstalledAppInfo>
)

@Serializable
data class StopGoalMessage(
    val type: String = "stop_goal"
)

@Serializable
data class VoiceStartMessage(
    val type: String = "voice_start"
)

@Serializable
data class VoiceChunkMessage(
    val type: String = "voice_chunk",
    val data: String
)

@Serializable
data class VoiceStopMessage(
    val type: String = "voice_stop",
    val action: String
)

@Serializable
data class ServerMessage(
    val type: String,
    val requestId: String? = null,
    val deviceId: String? = null,
    val message: String? = null,
    val sessionId: String? = null,
    val goal: String? = null,
    val success: Boolean? = null,
    val stepsUsed: Int? = null,
    val step: Int? = null,
    val action: JsonObject? = null,
    val reasoning: String? = null,
    val screenHash: String? = null,
    val x: Int? = null,
    val y: Int? = null,
    val x1: Int? = null,
    val y1: Int? = null,
    val x2: Int? = null,
    val y2: Int? = null,
    val duration: Int? = null,
    val text: String? = null,
    val packageName: String? = null,
    val url: String? = null,
    val code: Int? = null,
    // Intent fields
    val intentAction: String? = null,
    val intentUri: String? = null,
    val intentType: String? = null,
    val intentExtras: Map<String, String>? = null,
    val setting: String? = null
)
